torch/nn/parallel/distributed.py

```diff
@@ -320,3 +395,47 @@ def _process_batch():
         with torch.cuda.device(device_ids[0]):
             while True:
                 _process_batch()  # just to have a clear scope
+
+    @staticmethod
+    def _reduction_thread_fn_nccl(queue, group_id, device_ids, default_streams):
+        """
+        Reduction thread function specifically for NCCL, the reduction will use
+        default CUDA streams
+        """
+        def _process_batch():
+            dev_grad_batch, dev_events, job_event = queue.get()
+            dev_coalesced = []
+            # Coalesce the tensors on all devices and start a local reduction
+            for dev_id, grad_batch, event, default_stream in \
+                    zip(device_ids,
+                        dev_grad_batch,
+                        dev_events,
+                        default_streams):
+                with torch.cuda.device(dev_id), \
+                        torch.cuda.stream(default_stream):
+                    default_stream.wait_event(event)
+                    coalesced = _flatten_dense_tensors(grad_batch)
+                    dev_coalesced.append(coalesced)
+
+            # TODO: remove nccl.reduce with
+            #       dist.all_reduce_multigpus
+            nccl.reduce(dev_coalesced, root=0, streams=default_streams)
```

Comment:
This is confusing. You have already implemented the multigpu calls, so why not use the,?

Reply:
Will do it in a separate PR.  if we use allreduce_multigpu,  then user needs to use optimizer.step()  to update the gradient on each of the 8 ( or any number) GPUs instead of 1. In other words, passing each GPU's module's parameter to the optimizer and then do the step() for eight times.

This process is currently the performance bottleneck and I am still trying to figure out how to make it faster.  It would be great if you could comment on this as well

Reply:
that doesn't seem right - the parameters are broadcasted to all GPUs in `_sync_params`, which is why it's enough to use a single optimizer.

I'm worried about adding these new functions in these PR. They have no tests and are not used in here either. There's no way for us to figure out if they're correct.

Reply:
Once I move to dist. allreduce_multigpu, why do you still think we need to do broadcast parameters in _sync_params? Aren't we supposed to delete it? 

allreduce_multigpu should will let every single GPU to have the averaged gradients. And all the rest left to do at the beginning of each forward is to update the parameters on every GPU with the optimizer on 8 GPUs.

So a desired flow is that we don't broadcast anything in forward (maybe buffer?, but not parameters), and at the reduction thread, only use a single step of dist.all_reduce_multigpu(), which will use all 4 IB interfaces.  

For these new functions, I have my own script of testing it and I have already tested the functionality of the new functions.  As I put in the TODO in this PR, I will write the unittest for this too later.

Reply:
I know this is possible with NCCL, but this changes the public interface - the user would have to have separate optimizers for everything in `_module_copies`, and the copies are currently private. This is a breaking change of the interface.

If you don't want to unit test these functions, the please at least use them in Python code and check that it runs. Additionally, they look like ordinary non-experimental distributed functions, so please make them raise a warning saying that they are and shouldn't be used in user code (or just prefix them with an underscore).

Reply:
Yeah, that's why I didn't change the dist.all_reduce to dist.all_reduce_multigpu right now since it involves these interface changes and we later need to get some cleaner ways of doing it.

For the testing, yes, that's what I have already done in my own testing script, which is the python code that test all the new functions.  All the four functions work fine from my testing code.